feat(iOS) - blur filter using SwiftUI

intergalacticspacehighway opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

As per the discussion on the previous [PR thread](https://github.com/facebook/react-native/pull/52028#issuecomment-2979481948), this PR uses `SwiftUI` to implement blur filter on iOS.

## Approach:

To implement blur filter on iOS, we have two options:

1. Use `CAFilter` (private API, app can get rejected/API can break). Earlier [PR](https://github.com/facebook/react-native/pull/52028) was using that approach. Thanks to Nick for suggesting SwiftUI API.

2. Use `SwiftUI`. Wrap the view in a SwiftUI view and apply [blur](https://developer.apple.com/documentation/swiftui/view/blur(radius:opaque:)). This PR builds on top of that approach. This also enables a way to add `SwiftUI` only features like this one. Additional filters (grayscale, saturate, contrast, hueRotate) can also be added.

There are a few ways we can implement the SwiftUI approach:

1. Create a new `RCTSwiftUIComponentView` -> do style flattening in View -> check if `filter` is present and conditionally render the `RCTSwiftUIComponentView` on iOS, wrap children with a `SwiftUI` view. Tradeoff with this approach is that it adds `StyleSheet.flatten` overhead on JS side.
2. Add a `SwiftUI` container view inside of `RCTViewComponentView`. Tradeoff with this approach is that it complicates `RCTViewComponentView` a bit.


I decided to go with **2** to avoid the flattening tradeoff and try to minimize complicating `RCTViewComponentView`. it only adds the wrapper if it's required and removes if not (in this PR, blur filter style will add the wrapper, it will get removed if blur filter styling gets removed). It uses the existing container view pattern.

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Changelog:
[IOS][ADDED] - Filter blur

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

## Test Plan:

Test filter blur example on iOS. SwiftUI view should be added to the hierarchy.

<img src="https://github.com/user-attachments/assets/742539f4-a96d-45f4-94ba-5eb588d0ad5a" width="300px" />

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

## Aside:

- This PR also adds a new swift podspec. Creating a new podspec felt the right approach as adding swift in existing ones were adding some complexity. But open for changes here. Also, need some eyes on the podspec configs. cc - @chrfalch  🙏   this might also affect the SPM migration.
- Unrelated: Existing brightness filter has some inconsistency compared to android and web, it uses [self.layer.opacity](https://github.com/facebook/react-native/blob/6892dde36373bbef2d0afe535ae818b1a7164f08/packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm#L1008) so transparent background color do not blend well unless the view has an opacity. One solution would be to calculate true background color by using brightness or else use the `SwiftUI`'s [brightness](https://developer.apple.com/documentation/swiftui/view/brightness(_:)), which would be cleaner imo (tested and it works).

nandorojo commented:
Thanks for continuing to push us forward, @intergalacticspacehighway.

intergalacticspacehighway commented:
@cipolleschi i have been chatting with @NickGerleman regarding SwiftUI [here](https://github.com/facebook/react-native/pull/52028#issuecomment-2979481948), also we were having some conversation over discord. The intention of the PR was not to introduce official SwiftUI support, but it is more of an implementation detail. It is touching 2 core files, others are completely new files or small configs. I will look into SwiftPM migration, but i wanted to discuss the approach before investing too much time into it and Nick has some context on it, also I tagged @chrfalch for SwiftPM concern. Will make sure to do an RFC next time!

NickGerleman commented:
Sorry for the delay. Had a chance to go through all of this!

@joevilches will be a lot more familiar with how container view is set up. I know there was some intricacy, around how we set things up, depending on what we do with borders. I don't know if there is any way we can avoid the complexity with `transferVisualPropertiesFromView`, if we instead can add just another view to the hierarchy. For rare case where we have blur filters, we can make hierarchy larger I think. I think the added complexity to RCTViewComponentView is the thing that is spookiest with this change, that otherwise would be relatively safe, since we are not impacting existing view hierarchies.

A couple interesting scenarios came to mind, that would be helpful to validate explicitly:

1. Do we correctly handle adding and removing blur filter? I can see logic there
2. Is there any impact to things like hit testing within the filtered view?

The other bit, around complexity, is that we now have a "filter layer" with logic only ever used for brightness, where we get creative with blend modes to implement that. But the SwiftUI approach is really generic!

Afaict, SwiftUI has built in support for `hue-rotate`, `contrast`, `brightness`, `grayscale`, etc. So we could use this approach to fully replace the filter layer, and delete that added complexity, so we "trade away" the complexity of the current approach, for the more generic SwitftUI filter view that could support all of the different filters. We would need a feature flag for this, but I really like the idea of being able to support everything, and having only one path for filter effects, that doesn't rely on creativity with blending. I can see you were starting to add `grayscale` there, so I am guessing you were going that path for the others, so I think the only real ask here, is setting this up, so we could delete the old filter path, and keep things "simplier".

These all seem supported on iOS 13+ too, so we could fully support all the filters on iOS.

**Concrete Steps:** There are a couple interesting pieces of this. Can we split them up? E.g.

1. If we need any custom build logic, starting with that, since we now have both the build systems to deal with. We don't seem to need to branch out into Swift though, so maybe we could make this easy, and just keep things with the current component view?
2. Let's add a feature flag, to enable SwiftUI based filters. This can be generated, by adding to `ReactNativeFeatureFlags.config.js`, and running `yarn featureflags --update`.
3. Add the needed logic to use SwifUI container for brightness filters, instead of the current, gated in a way, where rolling out the SwifUI filters, would let us delete the current filter code
4. Add the rest of the filters?

intergalacticspacehighway commented:
Thank you for reviewing @NickGerleman!


> transferVisualPropertiesFromView, if we instead can add just another view to the hierarchy....

i thought about this, issue is when blur style gets added/removed, we'll need to clear the hierarchies, still adds some complexity. Also, we need to remove the visual styles from parent view's layer or else they make blur on child look weird 😓. So i thought simplest way would be copy the visual styles to SwiftUI content view layer. I also tried making a visual only layer and then just togging them between SwiftUI container and main container when blur gets added/removed. But this was adding more complexity.
The cleanest approach imo would be to use a new RCTViewComponentView inherited view which can abstract the layer logic so we don't touch the original RCTViewComponentView but that requires adding flatten style call in JS 😓 (i tried looking for a way to change views based on props but couldn't find an existing pattern, most of them do it in JS like multiline/singleline text input)

> Do we correctly handle adding and removing blur filter? I can see logic there

Yes, it mimics the existing container approach, so cleans it up when prop is not needed.

> Is there any impact to things like hit testing within the filtered view?

Yeah, there is. It should be fixed once we merge this - https://github.com/facebook/react-native/pull/52413. Will add example  in hit testing when blur/SwiftUI view is on.

> Do we have any examples with filters + transparent borders, which was IIRC the interesting container view case?

The usecase for container view iirc was overflow + boxShadow. i.e. even when overflow is hidden, boxShadow still needs to be out of bounds. This works with the blur since the new container becomes child of the SwiftUI content. Will add example to cover this.

> hue-rotate, contrast, brightness, grayscale

Yes! Once we finalize the approach, will do a separate PR to add them and replace the brightness filter and also the feature flag.

joevilches commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -793,44 +802,88 @@ - (BOOL)styleWouldClipOverflowInk
       ((!_props->boxShadow.empty() || (clipToPaddingBox && nonZeroBorderWidth)) || _props->outlineWidth != 0);
 }
 
+- (UIView *)childContainerView
```
name is kinda confusing, so is mine below. Can you be specific? This is only used for `filter` maybe change it to `filterView` or something

joevilches commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -793,44 +802,88 @@ - (BOOL)styleWouldClipOverflowInk
+  if (self.styleNeedsSwiftUIContainer) {
+    if (!_swiftUIWrapper) {
+      _swiftUIWrapper = [RCTSwiftUIContainerViewWrapper new];
+      _swiftUIWrapper.hostingView.frame = CGRectMake(0, 0, self.bounds.size.width, self.bounds.size.height);
+      UIView *swiftUIContentView =
+          [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.bounds.size.width, self.bounds.size.height)];
```
seems like the wrapper can handle this and we can encapsulate these 2 calls inside and just call some function the wrapper exposes that sets the frame/inits the content view

joevilches commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -575,6 +577,13 @@ - (void)updateLayoutMetrics:(const LayoutMetrics &)layoutMetrics
     self.layer.transform = RCTCATransform3DFromTransformMatrix(newTransform);
   }
+
+  if (_swiftUIWrapper) {
+    _swiftUIWrapper.hostingView.frame = self.bounds;
+    if (_swiftUIWrapper.contentView) {
+      _swiftUIWrapper.contentView.frame = self.bounds;
+    }
+  }
```
Can we encapsulate layout updates into the wrapper?

joevilches commented:
Thanks! This is so so awesome. I spent many days racking my brain trying to get this to work last year. Super excited to see you push through here!

I think the general approach is good, but I am trying to think of ways to reduce complexity, since the `containerView` I added a while back is quite confusing already, and now we are adding more "container views" for separate purposes. Figuring out where to add layers, subviews, etc is now quite complicated. For instance in the worst case you can have 4 subviews under a `RCTViewComponentView` before getting to the "actual" subviews - 2 views to support `filter`, 1 view to support clipping (`currentContainerView`), 1 view to support special content like on `Image` (`contentView`). 

Some ideas
- Removing the filter layer in the future. I see you mention that was planned so yay!
- I would try to encapsulate the fact that we need 2 UIViews to get this to work, that seems like a design detail. Ideally we remove those getters from the wrapper, and just have the wrapper expose the proper APIs need to insert this "View" somewhere in the hierarchy. I can imagine that would get a bit funky when you need to add stuff to the existing hierarchy, so I do not really think we can remove those APIs entirely, but its food for thought. Like I think anything layout related can be handled by the wrapper, since the hosting view and content view should be the same size. So the ask here would be to try and encapsulate it where it makes sense.
- Would also love if we did not have to do `transferVisualPropertiesFromView`, but realize that that is a bit challenging. I think ideally this all just sits *on top of* RCTViewComponentView when we have a `filter`, but I do not think we support that very well right now. Something in the mounting layer would have to change. But if we can do that then we remove a ton of complexity. I can ask around and see if this is something we can do.
- Not asking you to do this, it should be a separate PR, but: we really need some "content manager" or something to handle all the juggling. My brain hurts trying to figure out where everything goes. I think there is a future where everything we add to RCTViewComponentView goes through this manager and it adds the property to the right "content view", so we don't really need to think about if we missed some case by adding some layer/property/subview to the right place. cc @NickGerleman what do you think? I could try to tackle that during week when I am working on better engineering more.

joevilches reviewed (changes requested):
requesting changes for some of the things Nick asked for (like the feature flag), along with some of my nits, and an attempt to encapsulate some of the 2-view logic that is going on

intergalacticspacehighway commented:
Thank you @joevilches 🙏 

> But if we can do that then we remove a ton of complexity. I can ask around and see if this is something we can do.

This would be really amazing. We can make `RCTViewComponentView` simple again 🥲. I was just checking and seems `ShadowViewMutation::Create` has prop info and if we could pass it to `dequeueComponentViewWithComponentHandle` and return a new component descriptor based on it that could potentially solve it? We can remove the existing and the new Swift UI container to dedicated `RCTViewComponentView` inherited views.

Will make the above mentioned changes and update the PR. Thanks again!

thomazcapra commented:
👀

intergalacticspacehighway commented on packages/rn-tester/js/examples/Filter/FilterExample.js:
```diff
@@ -253,3 +281,35 @@ exports.examples = [
+    Animated.timing(animatedValue, {
+      toValue: isBlurred ? 0 : 20,
+      duration: 1000,
+      useNativeDriver: false,
```
For native driver support we need to merge this - https://github.com/facebook/react-native/pull/52920

intergalacticspacehighway commented:
Incorporated all the PR feedback and fixed SwiftPM integration. All checks passing now! lmk if anything needs to be changed  regarding naming/structure. I had to create two targets as SwiftPM does not work with mixing swift and obj-c. Also, the added feature flag is set to `false` for now. cc - @NickGerleman @cipolleschi @joevilches 🙏

intergalacticspacehighway commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -986,31 +1044,43 @@ - (void)invalidateLayer
-    _filterLayer.zPosition = CGFLOAT_MAX;
-    [self.layer addSublayer:_filterLayer];
+    
+    if (hasBrightnessFilter) {
```
Will make another PR to move brightness filter + add additional filters with SwiftUI.

AndrewPrifer commented:
Thanks for all the hard work @intergalacticspacehighway, can't wait for this to land!!

joevilches commented on packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm:
```diff
@@ -1410,6 +1480,65 @@ - (NSString *)componentViewName_DO_NOT_USE_THIS_IS_BROKEN
 }
 
+- (BOOL) styleNeedsSwiftUIContainer {
+  if (!_props->filter.empty()) {
```
for loop won't run below if this is false so I think we can just remove this if?

intergalacticspacehighway replied:
agree, pushed - https://github.com/facebook/react-native/pull/52495/commits/733338bc883c4fa099bed465bbcdf59b277934ec

joevilches reviewed (commented):
thanks for addressing some of the concerns and adding a feature flag!

joevilches commented:
@intergalacticspacehighway I am getting feedback with some colleagues on the general approach, as well as fighting some internal build system stuff to be happy with swift code, which is why there is an import but no subsequent merge. Will keep you in the loop

joevilches commented on packages/react-native/ReactApple/RCTSwiftUIWrapper/RCTSwiftUIContainerViewWrapper.m:
```diff
@@ -0,0 +1,49 @@
+
+#import "RCTSwiftUIContainerViewWrapper.h"
+
+@import RCTSwiftUI;
```
@intergalacticspacehighway thanks! verified that this breaks for me as well. any chance you can try and get it working with `<RCTSwiftUI/RCTSwiftUI-Swift.h>`? Everything else seems fine

intergalacticspacehighway replied:
@joevilches i tried it but it has quite a few cases. 

1. SPM requires manual search paths (brittle could break in future) with Swift.h syntax. I added a reply on this thread https://github.com/swiftlang/swift-package-manager/issues/7099#issuecomment-1815289087
2. The generated header path changes when we build with `USE_FRAMEWORKS=dynamic pod install`. We have this [conditional](https://github.com/expo/expo/blob/f59294f777fedfbfe3e4e67fd05187a95e98cd17/packages/expo/ios/Swift.h#L9) in expo.

So it seems `@import` is the only syntax without hacks. It also works with/without use frameworks. Maybe apple needs to update the guide.

We could've simply put the Swift and ObjC together with Cocoapods but for SPM support we need to put them into [different targets](https://developer.apple.com/documentation/xcode/build-settings-reference) when mixing objc and swift. 

> Targets can contain Swift, Objective-C/C++, or C/C++ code, but an individual target can’t mix Swift with C-family languages. 

Can you try the latest [commit](https://github.com/facebook/react-native/pull/52495/commits/cccf6b194a34ace31ac1e49e84e059d5f4c5462f) with the internal build system?

intergalacticspacehighway commented:
@joevilches Thank you so much for pushing this through! 🙏

@bigcupcoffee Nick has a [PR](https://github.com/facebook/react-native/pull/53794) to enable in nightly builds. We'll add an option to turn it on in Expo for experimental usage. Currently discussing it with the team.

joevilches commented:
@bigcupcoffee You can change how you init React Native and declare a release level, code pointers are 

* iOS: https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/AppDelegate/RCTReactNativeFactory.h#L96
* Android: https://github.com/facebook/react-native/blob/e1a6c38cc1d2268c8198d8acaccc048f583f4492/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/defaults/DefaultNewArchitectureEntryPoint.kt#L34

It is intentionally not well documented & tucked away due to how unstable these APIs can be at these levels, so use with caution! Please let me know if you see any bugs, crashes, or other issues.

bigcupcoffee commented:
Hey @joevilches yep this works great with 0.83 release thank you! So far I haven't seen much issues with it, is it possible to toggle individual flags in a similar fashion, or do I need to resort to patching, if I only want to enable swift ui filters so it'd be that tad bit safer? I'm really interested in adopting this feature (we have A LOT of usecases for this) and providing feedback to make it stable sooner 🙏

joevilches commented:
sorry been out sick for a bit :) thanks for the feedback thats great to hear! I have yet to find internal use case for this so far. I found a few instances of blurring in our code but it was for images, which works a bit differently and has designated support already. Perhaps there is a better way for us to gain some confidence here that is not manifesting an internal use case that may not exist. I can ask around.

No there is intentionally not a way to granularly control these flags without just changing them and patching like you are describing. The are obtuse by design since we prefer to hide that they exist due to their volatility. But I can understand your desire to change it and patch, I will try to see if we can expedite releasing this.

bigcupcoffee commented:
Hope you're feeling better! I'll probably go with a patch, it just would be nice to be able to take advantage of both this and prebuilt RN core. I don't have much native development experience but I'd suspect there's a chance I can just create my own override class and just invoke the override() myself in a similar fashion? Not sure if I can access those classes/namespaces but I'll give it a try and find out I guess. For now I've encountered some custom fonts issues with 0.83 that I'll report when I get the time for it, thanks again for helping!
